pkg/odo/cli/component/create.go

```diff
@@ -36,282 +43,298 @@ var (
-			componentPathType = occlient.GIT
-			checkFlag++
+func setCmpSourceAttrs(co *CreateOptions) (err error) {
+	componentCnt := 0
+
+	if len(co.componentBinary) != 0 {
+		if co.CreateArgs.SourcePath, err = util.GetAbsPath(co.componentBinary); err != nil {
+			return err
 		}
-		if len(componentLocal) != 0 {
-			componentPath = componentLocal
-			path, err := util.GetAbsPath(componentPath)
-			odoutil.LogErrorAndExit(err, "Failed to resolve %s to absolute path", componentPath)
-			componentPath = path
-			componentPathType = occlient.LOCAL
-			checkFlag++
+		co.CreateArgs.SourceType = occlient.BINARY
+		componentCnt++
+	}
+	if len(co.componentGit) != 0 {
+		co.CreateArgs.SourcePath = co.componentGit
+		co.CreateArgs.SourceType = occlient.GIT
+		componentCnt++
+	}
+	if len(co.componentLocal) != 0 {
+		if co.CreateArgs.SourcePath, err = util.GetAbsPath(co.componentLocal); err != nil {
+			return err
 		}
+		co.CreateArgs.SourceType = occlient.LOCAL
+		componentCnt++
+	}
 
-		if checkFlag > 1 {
-			log.Error("The source can be either --binary or --local or --git")
-			os.Exit(1)
-		}
+	if componentCnt > 1 {
+		return fmt.Errorf("The source can be either --binary or --local or --git")
+	}
 
-		// if --git is not specified but --ref is still given then error has to be thrown
-		if len(componentGit) == 0 && len(componentGitRef) != 0 {
-			log.Errorf("The --ref flag is only valid for --git flag")
-			os.Exit(1)
-		}
+	if len(co.componentGitRef) != 0 {
+		co.CreateArgs.SourceRef = co.componentGitRef
+	}
 
-		componentImageName, componentType, _, componentVersion := util.ParseCreateCmdArgs(args)
-
-		// Fetch list of existing components in-order to attempt generation of unique component name
-		componentList, err := component.List(client, applicationName)
-		odoutil.LogErrorAndExit(err, "")
-
-		// Generate unique name for component
-		componentName, err := component.GetDefaultComponentName(
-			componentPath,
-			componentPathType,
-			componentType,
-			componentList,
-		)
-		odoutil.LogErrorAndExit(err, "")
-
-		// Check to see if the catalog type actually exists
-		exists, err := catalog.Exists(client, componentType)
-		odoutil.LogErrorAndExit(err, "")
-		if !exists {
-			log.Errorf("Invalid component type: %v", componentType)
-			log.Info("Run 'odo catalog list components' for a list of supported component types")
-			os.Exit(1)
-		}
+	if len(co.componentGit) == 0 && len(co.componentGitRef) != 0 {
+		return fmt.Errorf("The --ref flag is only valid for --git flag")
+	}
 
-		// Check to see if that particular version exists
-		versionExists, err := catalog.VersionExists(client, componentType, componentVersion)
-		odoutil.LogErrorAndExit(err, "")
-		if !versionExists {
-			log.Errorf("Invalid component version: %v", componentVersion)
-			log.Info("Run 'odo catalog list components' to see a list of supported component type versions")
-			os.Exit(1)
-		}
+	return
+}
+
+func setCmpName(co *CreateOptions, args []string) (err error) {
```

Comment:
I would do `func (co *CreateOptions) setCmpName(args []string) (err error)` instead but that's just a matter of style :)

Reply:
Changed now :)